proto/wippersnapper/signal.proto

```diff
@@ -0,0 +1,104 @@
+import "error.proto";
+import "gps.proto";
+import "i2c.proto";
+import "i2c_output.proto";
+import "nanopb.proto";
+import "pixels.proto";
+import "pwm.proto";
+import "servo.proto";
+import "uart.proto";
+
+/*
+ * BrokerToDevice
+ * The BrokerToDevice message is sent from the broker to the device.
+ * It contains a oneof payload, which is a union of all the possible
+ * messages that can be sent from the broker to a device.
+ */
+message BrokerToDevice {
+  oneof payload {
+    // digitalio.proto
+    wippersnapper.digitalio.DigitalIOAdd digitalio_add       = 10;
+    wippersnapper.digitalio.DigitalIORemove digitalio_remove = 11;
+    wippersnapper.digitalio.DigitalIOEvent digitalio_event   = 12;
+    wippersnapper.digitalio.DigitalIOWrite digitalio_write   = 13;
+    // analogio.proto
+    wippersnapper.analogio.AnalogIOAdd analogio_add       = 20;
+    wippersnapper.analogio.AnalogIORemove analogio_remove = 21;
+    // checkin.proto
+    wippersnapper.checkin.CheckinResponse checkin_response = 30;
+    // servo.proto
+    wippersnapper.servo.ServoAdd servo_add       = 40;
+    wippersnapper.servo.ServoRemove servo_remove = 41;
+    wippersnapper.servo.ServoWrite servo_write   = 42;
+    // pwm.proto
+    wippersnapper.pwm.PWMAdd pwm_add                              = 50;
+    wippersnapper.pwm.PWMRemove pwm_remove                        = 51;
+    wippersnapper.pwm.PWMWriteDutyCycle pwm_write_duty            = 52;
+    wippersnapper.pwm.PWMWriteDutyCycleMulti pwm_write_duty_multi = 53;
+    wippersnapper.pwm.PWMWriteFrequency pwm_write_freq            = 54;
+    // pixels.proto
+    wippersnapper.pixels.PixelsAdd pixels_add       = 60;
+    wippersnapper.pixels.PixelsRemove pixels_remove = 61;
+    wippersnapper.pixels.PixelsWrite pixels_write   = 62;
+    // ds18x20.proto
+    wippersnapper.ds18x20.Ds18x20Add ds18x20_add       = 70;
+    wippersnapper.ds18x20.Ds18x20Remove ds18x20_remove = 71;
+    // uart.proto
+    wippersnapper.uart.UartAdd uart_add       = 80;
+    wippersnapper.uart.UartRemove uart_remove = 81;
+    wippersnapper.uart.UartWrite uart_write = 82;
+    // i2c.proto
+    wippersnapper.i2c.I2cBusScan i2c_bus_scan                      = 90;
+    wippersnapper.i2c.I2cDeviceAddOrReplace i2c_device_add_replace = 91;
+    wippersnapper.i2c.I2cDeviceRemove i2c_device_remove            = 92;
+    wippersnapper.i2c.I2cDeviceOutputWrite i2c_device_output_write = 93;
+    // error.proto
+    wippersnapper.error.Error error = 100;
+  }
+}
+
+/*
+ * DeviceToBroker
+ * The DeviceToBroker message is sent from the device to the broker.
+ * It contains a oneof payload, which is a union of all the possible
+ * messages that can be sent from a device to the broker.
+ */
+message DeviceToBroker {
+  option (nanopb_msgopt).submsg_callback = true;
+  oneof payload {
+    //digitalio.proto
+    wippersnapper.digitalio.DigitalIOEvent digitalio_event = 10;
+    // analogio.proto
+    wippersnapper.analogio.AnalogIOEvent analogio_event = 20;
```

Comment:
are we intentionally skipping the "added" responses for digital and analog?

Reply:
the APIs don’t attach or create anything, its internal to the device so they are excluded from the pb

but if we are going to refactor to have the broker ensure the device got the message and did something then yeah I think we should add them